src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ConnectHelper.cs

```diff
@@ -15,6 +17,125 @@ namespace System.Net.Http
 {
     internal static class ConnectHelper
     {
+        public static ValueTask<Stream> ConnectAsync(string host, int port, bool async, CancellationToken cancellationToken)
+        {
+            return async ? ConnectAsync(host, port, cancellationToken) : new ValueTask<Stream>(Connect(host, port, cancellationToken));
+        }
+
+        private static async ValueTask<Stream> ConnectAsync(string host, int port, CancellationToken cancellationToken)
+        {
+            // We use the static Socket.ConnectAsync with a SocketAsyncEventArgs, because this approach is:
+            // 1. Cancellable
+            // 2. Does not create Dual-stack sockets, which are unavailable in certain environments,
+            //    see https://github.com/dotnet/runtime/issues/44686.
+            var saea = new ConnectEventArgs();
+            try
+            {
+                saea.Initialize(cancellationToken);
+
+                // Configure which server to which to connect.
+                saea.RemoteEndPoint = new DnsEndPoint(host, port);
+
+                // Initiate the connection.
+                if (Socket.ConnectAsync(SocketType.Stream, ProtocolType.Tcp, saea))
+                {
+                    // Connect completing asynchronously. Enable it to be canceled and wait for it.
+                    using (cancellationToken.UnsafeRegister(static s => Socket.CancelConnectAsync((SocketAsyncEventArgs)s!), saea))
+                    {
+                        await saea.Builder.Task.ConfigureAwait(false);
+                    }
+                }
+                else if (saea.SocketError != SocketError.Success)
+                {
+                    // Connect completed synchronously but unsuccessfully.
+                    throw new SocketException((int)saea.SocketError);
+                }
+
+                Debug.Assert(saea.SocketError == SocketError.Success, $"Expected Success, got {saea.SocketError}.");
+                Debug.Assert(saea.ConnectSocket != null, "Expected non-null socket");
+
+                // Configure the socket and return a stream for it.
+                Socket socket = saea.ConnectSocket;
+                socket.NoDelay = true;
+                return new NetworkStream(socket, ownsSocket: true);
+            }
+            catch (Exception error) when (!(error is OperationCanceledException))
+            {
+                throw CreateWrappedException(error, host, port, cancellationToken);
+            }
+            finally
+            {
+                saea.Dispose();
+            }
+        }
+
+        private static Stream Connect(string host, int port, CancellationToken cancellationToken)
+        {
+            // For synchronous connections, we can just create a socket and make the connection.
+            cancellationToken.ThrowIfCancellationRequested();
+            var socket = new Socket(SocketType.Stream, ProtocolType.Tcp);
```

Comment:
Isn't this creating a dual-mode socket?

Reply:
Yes, but there was no sync in 3.1. I dont like this inconsistence either, just took the state at the commit I mentioned.

The bad thing is that the ony alternatives I see is to either do sync over async or implement a sync version of [DnsConnect](https://github.com/dotnet/runtime/blob/34cae7461f96a9c1a7574c72c436e309bf4031f4/src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.cs#L655) within System.Net.Http